Clients/Xamarin.Interactive.MTouch/MTouchSdkTool.cs

```diff
@@ -46,40 +47,80 @@ public static string GetMlaunchPath ()
             throw new MlaunchNotFoundException ();
         }
 
-        static string RunTool (string fileName, string arguments)
+        static Task<string> RunToolAsync (string fileName, string arguments, int timeout = 5000)
         {
+            var tcs = new TaskCompletionSource<string> ();
+
             var proc = new Process {
                 StartInfo = new ProcessStartInfo {
                     FileName = fileName,
                     Arguments = arguments,
                     UseShellExecute = false,
                     RedirectStandardOutput = true,
                     RedirectStandardError = true
-                }
+                },
+                EnableRaisingEvents = true
             };
 
+            proc.Exited += (o, args) => {
+                if (proc.ExitCode == 0) {
+                    try {
+                        tcs.TrySetResult (proc.StandardOutput.ReadToEnd ());
+                    } catch (Exception e) {
+                        tcs.TrySetException (e);
+                    }
+                } else
+                    tcs.TrySetException (new Exception ($"'{fileName} {arguments}' exited with exit code {proc.ExitCode}"));
+            };
+
+            Log.Info (TAG, $"{fileName} {arguments}");
             proc.Start ();
-            proc.WaitForExit ();
 
-            if (proc.ExitCode == 0)
-                return proc.StandardOutput.ReadToEnd ();
+            if (timeout > 0) {
+                Task.Run (() =>  {
+                    if (!proc.WaitForExit (timeout)) {
+                        Log.Info (TAG, $"TIMEOUT {fileName} {arguments}");
+                        tcs.TrySetException (new TimeoutException ());
+                        proc.Kill ();
+                    }
+                });
+            }
+
+            return tcs.Task;
+        }
 
-            throw new Exception ($"'{fileName} {arguments}' exited with exit code {proc.ExitCode}");
+        static async Task<string> RunToolWithRetriesAsync (
+            string fileName,
+            string arguments,
+            int timeoutRetries = 3,
+            int timeout = 5000)
+        {
+            for (var i = 0; i < timeoutRetries; i++) {
+                try {
+                    return await RunToolAsync (fileName, arguments, timeout);
```

Comment:
Exponential backoff please. This allows you to not care about the actual timeout duration since you compute it from the number of iterations you're willing to attempt. Generally provides a much smoother user experience as well with more attempts within the same overall window.

To fit within the 15 seconds you've got here, you can do 7 attempts with a 100×2ⁱ ms backoff (we use this elsewhere), which would be 12 seconds total wait across 7 attempts:

```csharp
… timeoutRetries = 7 …

for (int i = 0; i < timeoutRetries; i++) {
    try {
        return await RunToolAsync (
            fileName,
            arguments,
            (int)Math.Pow (2, i) * 100));

        …
}
```


![backoff](https://user-images.githubusercontent.com/49539/40335002-5c0654be-5d2f-11e8-9a9c-c13fa4f3e3b0.gif)

Reply:
Isn't exponential backoff more about time _between_ retries? The caller here still knows best what kind of timeout is appropriate for the given process (mlaunch can be slow, for example).

So given that we do care about the actual timeout duration, I'm not sure what a backoff approach buys us. This isn't anything like network congestion.

Reply:
Never mind. It was not clear how this timeout was actually being used. So just disregard the whole bit.